:new: [policies] Add support to apply constructible policy checks on …

kris-jusiak opened the pull request:
…the root type

Problem:
- There is no way to apply constructible policy on types being created (root types - `create<RootType>`)

Solution:
- Expose `is_root` which allows to check whether the policy check is called for the `RootType`
- Add `constructible<true>` which will trigger the policy checks for the `RootType`

kris-jusiak committed:
:new: [policies] Add support to apply constructible policy checks on the root type

Problem:
- There is no way to apply constructible policy on types being created (root types - `create<RootType>`)

Solution:
- Expose `is_root` which allows to check whether the policy check is called for the `RootType`
- Add `constructible<true>` which will trigger the policy checks for the `RootType`

kris-jusiak commented:
@kanstantsin-chernik

kanstantsin-chernik commented:
@krzysztof-jusiak looks good. Tomorrow will try to integrate and share feedback.

kanstantsin-chernik commented:
@krzysztof-jusiak did you see that test coverage had been decreased?

kanstantsin-chernik commented on test/ft/di_config.cpp:
```diff
@@ -68,6 +68,36 @@ test constructible_policy_must_be_bound = [] {
+      const auto is_bound_root_shared_ptr = is_root<di::_>{} && is_shared_ptr<di::_>{} && is_bound<di::_>{};
+      const auto is_bound_not_root = !is_root<di::_>{} && is_bound<di::_>{};
+
+      return di::make_policies(constructible<true>(is_bound_root_shared_ptr || is_bound_not_root));
```
@krzysztof-jusiak Being hones I don't think this `<true>` is self-explained. It would be better to have anything like:
`constructible_including_root` and `constructible_excluding_root` or just have one and if there is is_root then we differenctiate if no then only for non-root objects. I didn't noticed this `<true>` first time and was quite surprised that `is_root` doesn't work

kris-jusiak replied:
Yeah, I started with that but couldn't find a good name ;) so I let the user decide. It's a limitation of the C++ since we can't do IncludeRoot = true, but user can specify any name they want and just use it instead of true/false

kanstantsin-chernik replied:
@krzysztof-jusiak What about having just argument is_root_included? It might be more explicit though

kanstantsin-chernik commented on test/ft/di_config.cpp:
```diff
@@ -68,6 +68,36 @@ test constructible_policy_must_be_bound = [] {
+      using namespace di::policies::operators;
+
+      const auto is_bound_root_shared_ptr = is_root<di::_>{} && is_shared_ptr<di::_>{} && is_bound<di::_>{};
+      const auto is_bound_not_root = !is_root<di::_>{} && is_bound<di::_>{};
```
This does work for my project but generates warning only. How can we make it error?

kris-jusiak replied:
It's always a warning, I think. You don't compile with `-Werror`?

kanstantsin-chernik replied:
Unfortunately no :) may be one day...

kris-jusiak replied:
hmm, did you have errors before? I didn't change the reporting it should be a warning and a linker error (it's a trick to avoid long error messages showing all call stack)

kanstantsin-chernik replied:
Yes, I am getting tons of errors but I was surprised this time it was one warning

kris-jusiak replied:
hmm, interesting, I'll check in msvc

kanstantsin-chernik commented:
@krzysztof-jusiak I've added few inline comments.

kris-jusiak commented:
Regarding the coverage, though. Yes, I've noticed it but I think it's a false positive because all path are tested (I was doing it in TDD) and not sure what's codecov is really complaining about?

kanstantsin-chernik commented:
It seems we can check why codecov decreased coverage: https://docs.codecov.io/docs/unexpected-coverage-changes#section-review-reports-yourself

kris-jusiak committed:
:wrench: Add .gitignore

Problem:
- build directory is an artefact of building

Solution:
- Add .gitignore and ignore the build directory
